File: src/scheduler.ts
import type { Ranker } from "./model";
import type { RankRequest } from "./protocol";

type Job<Result> = {
  peer: string;
  request: RankRequest;
  readyAt: number;
  expiresAt: number;
  reply: (best: Result | 1, status: "ok" | "fallback") => void;
};

/** One inference at a time, with at most the latest pending request per peer. */
export class Scheduler<Result = number> {
  private pending = new Map<string, Job<Result>>();
  private active?: { job: Job<Result>; controller: AbortController };
  private timer?: ReturnType<typeof setTimeout>;
  private closed = false;

  constructor(
    private rank: Ranker<Result>,
    private options: {
      debounceMs: number;
      timeoutMs: number;
      correctionTimeoutMs?: number;
      correctionDebounceMs?: number;
    } = { debounceMs: 80, timeoutMs: 1200 },
  ) {}

  submit(peer: string, request: RankRequest, reply: Job<Result>["reply"]) {
    if (this.closed) return;
    if (this.active?.job.peer === peer) this.active.controller.abort();
    if (!this.pending.has(peer) && this.pending.size >= 32) {
      reply(1, "fallback");
      return;
    }
    const now = performance.now();
    const correction = request.pinyin !== undefined;
    this.pending.set(peer, {
      peer,
      request,
      reply,
      readyAt:
        now +
        (correction
          ? (this.options.correctionDebounceMs ?? 240)
          : this.options.debounceMs),
      expiresAt:
        now +
        (correction
          ? (this.options.correctionTimeoutMs ?? 3000)
          : this.options.timeoutMs),
    });
    this.pump();
  }

  close() {
    this.closed = true;
    if (this.timer) clearTimeout(this.timer);
    this.pending.clear();
    this.active?.controller.abort();
  }

  private pump() {
    if (this.closed || this.active) return;
    if (this.timer) clearTimeout(this.timer);
    const job = [...this.pending.values()].sort(
      (a, b) => a.readyAt - b.readyAt,
    )[0];
    if (!job) return;
    const wait = job.readyAt - performance.now();
    if (wait > 0) {
      this.timer = setTimeout(() => this.pump(), wait);
      return;
    }
    this.pending.delete(job.peer);
    if (job.expiresAt <= performance.now()) {
      job.reply(1, "fallback");
      this.pump();
      return;
    }
    const controller = new AbortController();
    this.active = { job, controller };
    const timeout = setTimeout(
      () => controller.abort(),
      job.expiresAt - performance.now(),
    );
    void this.run(job, controller).finally(() => {
      clearTimeout(timeout);
      this.active = undefined;
      this.pump();
    });
  }

  private async run(job: Job<Result>, controller: AbortController) {
    const interrupted = Promise.withResolvers<never>();
    const onAbort = () => interrupted.reject(controller.signal.reason);
    controller.signal.addEventListener("abort", onAbort, { once: true });
    try {
      // Abort is advisory to the ranker. Bound our own wait even if a transport
      // never settles its promise after cancellation, so the queue can recover.
      if (controller.signal.aborted) onAbort();
      const best = await Promise.race([
        this.rank(job.request, controller.signal),
        interrupted.promise,
      ]);
      if (!controller.signal.aborted && !this.closed) job.reply(best, "ok");
      else if (!this.closed && !this.pending.has(job.peer))
        job.reply(1, "fallback");
    } catch {
      if (!this.closed && !this.pending.has(job.peer)) job.reply(1, "fallback");
    } finally {
      controller.signal.removeEventListener("abort", onAbort);
    }
  }
}

File: src/riwen/scheduler.py
"""Serial inference with per-peer replacement and end-to-end deadlines."""
import asyncio
from dataclasses import dataclass


@dataclass
class Job:
    peer: object
    request: object
    reply: object
    ready: float
    expires: float


def consume(task):
    if not task.cancelled():
        task.exception()


class Scheduler:
    def __init__(self, rank, debounce=0.08, timeout=1.2, correction_debounce=0.24, correction_timeout=3):
        self.rank = rank
        self.debounce, self.timeout = debounce, timeout
        self.correction_debounce, self.correction_timeout = correction_debounce, correction_timeout
        self.pending = {}
        self.active = None
        self.interrupted = None
        self.timer = None
        self.runner = None
        self.closed = False

    def submit(self, peer, request, reply):
        if self.closed:
            return
        if self.active and self.active.peer == peer:
            self.interrupted.set()
        if peer not in self.pending and len(self.pending) >= 32:
            reply(1, "fallback")
            return
        now = asyncio.get_running_loop().time()
        correction = request.pinyin is not None
        self.pending[peer] = Job(peer, request, reply,
                                 now + (self.correction_debounce if correction else self.debounce),
                                 now + (self.correction_timeout if correction else self.timeout))
        self.pump()

    def pump(self):
        if self.closed or self.active:
            return
        if self.timer:
            self.timer.cancel()
            self.timer = None
        loop = asyncio.get_running_loop()
        while self.pending:
            job = min(self.pending.values(), key=lambda item: item.ready)
            if job.ready > loop.time():
                self.timer = loop.call_at(job.ready, self.pump)
                return
            del self.pending[job.peer]
            if job.expires <= loop.time():
                job.reply(1, "fallback")
                continue
            self.active = job
            self.interrupted = asyncio.Event()
            self.runner = asyncio.create_task(self.run(job, self.interrupted))
            return

    async def run(self, job, interrupted):
        inference = asyncio.create_task(self.rank(job.request))
        inference.add_done_callback(consume)
        abort = asyncio.create_task(interrupted.wait())
        try:
            done, _ = await asyncio.wait((inference, abort),
                                         timeout=max(0, job.expires - asyncio.get_running_loop().time()),
                                         return_when=asyncio.FIRST_COMPLETED)
            if (inference in done and not interrupted.is_set() and not self.closed
                    and asyncio.get_running_loop().time() < job.expires):
                job.reply(inference.result(), "ok")
            elif not self.closed and job.peer not in self.pending:
                job.reply(1, "fallback")
        except (Exception, asyncio.CancelledError):
            if not self.closed and job.peer not in self.pending:
                job.reply(1, "fallback")
        finally:
            # Do not await a ranker that ignores cancellation. Our own deadline
            # releases the queue; the actual HTTP client closes its socket.
            inference.cancel()
            abort.cancel()
            self.active = None
            self.pump()

    def close(self):
        self.closed = True
        self.pending.clear()
        if self.timer:
            self.timer.cancel()
        if self.interrupted:
            self.interrupted.set()
